### PR TITLE
[v2] Add the toc function with the test failure.

### PR DESCRIPTION
**Pull request by JiaxiangBU**
This follows the original [pull request](https://github.com/jdblischak/workflowr/pull/154).
I add my original function in the `dev` branch and pull it into your `dev` branch, just test the git process you show me.
And I will follow your tips to edit the function in the two parts,

1. relative path
1. change the base logic of the function

```
s <- wflow_status()
rmd <- rownames(s$status)[s$status$published]
html <- to_html(rmd, outdir = s$docs)
html <- fs::path_file(html)
```

Just let you know, thank you for your help, I learn a lot from you.

### CODE REVIEW

**Comment by jdblischak** [comment]
> I add my original function in the dev branch and pull it into your dev branch, just test the git process you show me.

@JiaxiangBU Great!

> change the base logic of the function

This will be the biggest change. By using workflowr-specific functions, you should be able to write the function without adding any new dependencies. The only new dependency you should need to add is clipr.

> Just let you know, thank you for your help, I learn a lot from you.

Of course! Thanks for helping improve workflowr!

Please let me know when you are ready for me to review the pull request.

**Comment by JiaxiangBU** [comment]
@jdblischak I follows your tips to use 

1. `to_html` to get HTML paths
1. `rmarkdown::yaml_front_matter` to get titles
1. relative paths

and pull it, just let you know.

File: R/wflow_toc.R [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#' @import purrr
+#' @import readr
+#' @import tidyr
+#' @import clipr
+#' @import here
+#' @import rmarkdown
+#' @import tibble
+#' @export
+
+wflow_toc <- function(analysis_dir = "analysis", docs_dir = "docs") {
```
Comment (jdblischak) [inline R/wflow_toc.R]: The only argument the function needs is `project = "."`. `wflow_status()` returns the paths to these subdirectories.

File: R/wflow_toc.R [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  s <- wflow_status()
```
Comment (jdblischak) [inline R/wflow_toc.R]: After adding `project = "."` as the default argument to `wflow_toc()`, you can change `wflow_status()` to `wflow_status(project = project)`.

File: R/wflow_toc.R [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  s$status %>%
+    tibble::rownames_to_column(var = 'rmd_path') %>%
```
Comment (jdblischak) [inline R/wflow_toc.R]: Do you know how to perform these steps without relying on the tidyverse functions? My package doesn't currently use these packages, and I don't want to add a lot of new dependencies for just this one function. If needed, I can translate these to base R commands.

File: R/wflow_toc.R [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    tidyr::unnest() %>%
+
+    dplyr::mutate(
+      name = ifelse(is.na(name), html_path, name)
+    ) %>%
+    dplyr::mutate(
+      link = glue::glue("1. [{name}]({html_path})")
+    ) %>%
+    .$link %>%
+    clipr::write_clip()
```
Comment (jdblischak) [inline R/wflow_toc.R]: I'd like to make clipr a suggested package. If the user doesn't have clipr installed, then the function can print the results to the R console. Either way it would be good to return the character vector invisibly using `return(invisible(x))`, where `x` would be replaced with the character vector containing the toc.

You can check if a package is installed using the following strategy:

```
if requireNamespace("clipr", quietly = TRUE) {
  # Write to clipboard
} else {
  # Print to R console
}
```

**Review by jdblischak — CHANGES_REQUESTED** [review]
This is coming along great! Please see my comments and let me know what questions you have.

File: R/wflow_toc.R (hunk 1 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: My bad, I forget it, I will change it in the next commit.

File: R/wflow_toc.R (hunk 2 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: LGTM, I will do it.

File: R/wflow_toc.R (hunk 3 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: Yes, I agree with you. I will try it first.

Comment (JiaxiangBU) [reply R/wflow_toc.R]: Yes, I agree with you. I will try it first.

File: R/wflow_toc.R (hunk 1 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I edit it, here, https://github.com/jdblischak/workflowr/pull/155/commits/8787eedb8db24ed51620f206765db6c2985e280f

File: R/wflow_toc.R (hunk 2 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I edit it, here, https://github.com/jdblischak/workflowr/pull/155/commits/8787eedb8db24ed51620f206765db6c2985e280f

File: R/wflow_toc.R (hunk 3 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I edit it, but I don't find a good substitute for `dplyr::*join` function, do you have any recommendations?
here is my handle, https://github.com/jdblischak/workflowr/pull/155/commits/cdbb9c3fcde89164564e6002b97bc93d37f1563a

File: R/wflow_toc.R (hunk 4 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I use your recommendation, but I do it in my way because I still get the point of the function `invisible`.
In my understanding, if the users have already installed the package `clipr`, they will give the toc in their clipboard. If not, the function `wflow_toc` will print the toc in the console?

Do I miss something?

**Comment by JiaxiangBU** [comment]
Hi @jdblischak , I think I finish most of them, and I am still working on some minimal parts. Now I am trying to update the `DESC` manually, do you know to have to update it automatically?
In our conversation, we will delete most of the dependencies from the package `tidyverse`.

**Comment by JiaxiangBU** [comment]
Hi @jdblischak , I think the rmarkdown chunk option `results='asis'` is better idea.

> when set to `'asis'`, text output is written “as-is”, e.g., you can write out raw Markdown text from R code (like `cat('**Markdown** is cool.\n'))`

But I don't find a way to set this argument inside the function. Do you know something about it?
Here is some discussion on [SO](https://stackoverflow.com/questions/52631689/how-to-show-formatted-r-output-with-results-asis-in-rmarkdown)

**Comment by jdblischak** [comment]
>  Now I am trying to update the DESC manually, do you know to have to update it automatically?

@JiaxiangBU To update `DESCRIPTION` automatically, first delete the `@import` statements from the roxygen2 documentation and then run `devtools::document()`.

**Comment by JiaxiangBU** [comment]
> > Now I am trying to update the DESC manually, do you know to have to update it automatically?
> 
> @JiaxiangBU To update `DESCRIPTION` automatically, first delete the `@import` statements from the roxygen2 documentation and then run `devtools::document()`.

Hi, @jdblischak I do it and drop most of the tidyverse related packages.

**Comment by jdblischak** [comment]
@JiaxiangBU My mistake. `document()` only automatically updates `NAMESPACE`. You'll need to manually edit `DESCRIPTION` to remove the packages. When you do that, please add back the version specifications for fs and rmarkdown.

File: R/wflow_toc.R [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#' @import rmarkdown
+#' @export
+
+wflow_toc <- function(project = ".") {
+  s <- wflow_status(project = ".")
+  df<- s$status
+  df$rmd_path <- rownames(df)
+  df <- df[df$published]
+  df$html_path <- to_html(df$rmd_path,outdir = s$docs) %>% basename()
+  file_name <- sapply(df$rmd_path, function(x) rmarkdown::yaml_front_matter(x)$title) %>%
```
Comment (jdblischak) [inline R/wflow_toc.R]: This part is tricky since an Rmd file may not have a title defined. I think it may be best handled with a custom function. Can I send push a commit to your pull request to add this? Then you would get the change locally by running `git pull`.

File: R/wflow_toc.R [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    na.omit() %>%
+    as.data.frame() %>%
+    `names<-`('name')
+  file_name$rmd_path <- rownames(file_name)
+  df <- df %>% dplyr::inner_join(file_name, by = 'rmd_path')
+  df$name <- ifelse(is.na(df$name), df$html_path, as.character(df$name))
+  df$link <- paste0("1. [",df$name,"](",df$html_path,")")
+
+  # output
+  # if (requireNamespace("clipr", quietly = TRUE)) {
```
Comment (jdblischak) [inline R/wflow_toc.R]: To be able to view the output even though it is copied to the clipboard, you can return the value of `df$link` at the end of the function with `return(invisible(df$link))`. 

```
> x <- wflow_toc()
The table of content of your project is on the clipboard.
> x
[1] "1. [About](about.html)"     "1. [Home](index.html)"      "1. [License](license.html)"
```

This will also make it easier to write automated tests for `wflow_toc()`.

**Review by jdblischak — CHANGES_REQUESTED** [review]
This is coming along great!

File: R/wflow_toc.R [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#' @import dplyr
+#' @import clipr
+#' @import rmarkdown
+#' @export
+
+wflow_toc <- function(project = ".") {
+  s <- wflow_status(project = ".")
+  df<- s$status
+  df$rmd_path <- rownames(df)
+  df <- df[df$published]
```
[the remark this answers is not in the document]
Comment (JiaxiangBU) [reply R/wflow_toc.R]: Yes, I agree with you.

[the remark this answers is not in the document]
Comment (JiaxiangBU) [reply R/wflow_toc.R]: Yes, I agree with you.

File: R/wflow_toc.R (hunk 5 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: Sure, that is great, I am looking forward to it.

File: R/wflow_toc.R (hunk 6 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I get your point, so the users expect to get an output in a vector.

**Comment by JiaxiangBU** [comment]
> @JiaxiangBU My mistake. `document()` only automatically updates `NAMESPACE`. You'll need to manually edit `DESCRIPTION` to remove the packages. When you do that, please add back the version specifications for fs and rmarkdown.

@jdblischak 
Yeah, I delete the tidyverse packages in the DESC document and put the `clipr` in the suggested field.

File: R/wflow_toc.R (hunk 5 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I try a pull request from your `dev` branch to mine and now I know how it works.

File: R/wflow_toc.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I make a new commit for this editing.

File: R/wflow_toc.R (hunk 6 above)
Comment (JiaxiangBU) [reply R/wflow_toc.R]: I make a new commit to edit it.

**Comment by jdblischak** [comment]
@JiaxiangBU Thanks for the recent round of updates! I've made some changes that I'd like to push to your branch, but I'm having trouble doing so. Can you please make sure to check the box "Allow edits from maintainers" and then let me know once it is enabled?

https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

In the meantime, please don't make any additional changes to your Pull Request. Thanks!

**Comment by JiaxiangBU** [comment]
> @JiaxiangBU Thanks for the recent round of updates! I've made some changes that I'd like to push to your branch, but I'm having trouble doing so. Can you please make sure to check the box "Allow edits from maintainers" and then let me know once it is enabled?
> 
> https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork
> 
> In the meantime, please don't make any additional changes to your Pull Request. Thanks!

Hi @jdblischak , I double check it, it still stays in the `"Allow edits from maintainers"` status.
image

**Comment by jdblischak** [comment]
> I double check it, it still stays in the "Allow edits from maintainers" status.

@JiaxiangBU Thanks for the confirmation! The problem was on my end. For some reason Git wasn't allowing me to used my preferred approach (pushing from a detached HEAD state). I pushed my changes.

Next topic for discussion. Does it make sense to include `index.html` in the table of contents?

**Comment by jdblischak** [comment]
Also, please pull my recent commits to your local repository by running `git pull origin dev`

**Comment by JiaxiangBU** [comment]
> Also, please pull my recent commits to your local repository by running `git pull origin dev`

Yes, I think I successfully pull your lastest editing in my local branch.

```
$ git pull origin dev
remote: Enumerating objects: 42, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 30 (delta 23), reused 30 (delta 23), pack-reused 0
Unpacking objects: 100% (30/30), done.
From https://github.com/JiaxiangBU/workflowr
 * branch            dev        -> FETCH_HEAD
   54349e9..1adaf36  dev        -> origin/dev
Updating 54349e9..1adaf36
Fast-forward
 DESCRIPTION                     |  7 +++---
 NAMESPACE                       |  2 --
 R/wflow_toc.R                   | 65 ++++++++++++++++++++++++++++--------------------------
 _pkgdown.yaml                   |  1 +
 inst/WORDLIST                   |  1 +
 man/wflow_toc.Rd                | 18 ++++++++-------
 tests/testthat/test-wflow_toc.R | 34 ++++++++++++++++++++++++++++
 7 files changed, 83 insertions(+), 45 deletions(-)
 create mode 100644 tests/testthat/test-wflow_toc.R
```

I find the function `wflow_toc` is the same as one in your `dev` branch.

**Comment by JiaxiangBU** [comment]
> Next topic for discussion. Does it make sense to include `index.html` in the table of contents?

Yes, currently I use it in the `index.html` pages for several workflowr projects.

**Comment by jdblischak** [comment]
> Yes, currently I use it in the index.html pages for several workflowr projects.

@JiaxiangBU My question was about including `index.html` in the list of analyses.

For example, on your main page for [learn_rmd](https://jiaxiangbu.github.io/learn_rmd/), your list of analyses includes `Home`, `About`, and `License`. Since links to these pages are already available in the navigation bar, it is somewhat redundant to include them in the list of rmd files. We could by default have `wflow_toc()` ignore any rmd files that are included in the navigation bar (by checking `analysis/_site.yml`), and then provide a function argument to optionally include them. Does that make sense?

**Comment by JiaxiangBU** [comment]
> > Yes, currently I use it in the index.html pages for several workflowr projects.
> 
> @JiaxiangBU My question was about including `index.html` in the list of analyses.
> 
> For example, on your main page for [learn_rmd](https://jiaxiangbu.github.io/learn_rmd/), your list of analyses includes `Home`, `About`, and `License`. Since links to these pages are already available in the navigation bar, it is somewhat redundant to include them in the list of rmd files. We could by default have `wflow_toc()` ignore any rmd files that are included in the navigation bar (by checking `analysis/_site.yml`), and then provide a function argument to optionally include them. Does that make sense?

 I get your point and agree with you. I am thinking about it. When I update the function, I will let you know,.

**Comment by JiaxiangBU** [comment]
Next topic, I find based on this function, the toc is updated manually by users every time. Maybe an automatic solution is better.

**Comment by JiaxiangBU** [comment]
Hi John. I make a new commit for this.

1. title and HTML are extracted from this object, `rmd`, so I edit it first.
1. And I find the imported packages in the function document is deleted, so I don't add here package.

**Comment by jdblischak** [comment]
> I make a new commit for this.

@JiaxiangBU Awesome! Good job parsing the YAML file!

> title and HTML are extracted from this object, rmd, so I edit it first.

I agree that you want to filter the results early.

> And I find the imported packages in the function document is deleted, so I don't add here package.

You don't need to use the here package. `wflow_status()` returns the path to the analysis directory (note that it doesn't need to called `analysis/`; a user can change it and `wflow_status()` will still identify the correct subdirectory).

I'll update the tests.

**Comment by jdblischak** [comment]
@JiaxiangBU I just pushed two new commits to your branch. Please pull them locally with `git pull origin dev`.

I changed the filtering to use `html`. This is because these are always just the filenames like in the navigation bar. The `rmd` vector may have other parts of the file path included, which is why it won't always work (we have to preserve the path to be able to read the title later on).

I also added some tests to catch some edge cases, e.g. when there is no navbar.

> Next topic, I find based on this function, the toc is updated manually by users every time. Maybe an automatic solution is better.

Could you please elaborate on this idea? What would happen automatically?

**Comment by JiaxiangBU** [comment]
> You don't need to use the here package. `wflow_status()` returns the path to the analysis directory (note that it doesn't need to called `analysis/`; a user can change it and `wflow_status()` will still identify the correct subdirectory).

Thanks, I have a deep understanding of this functon `wflow_status()`.

**Comment by JiaxiangBU** [comment]
> I changed the filtering to use `html`. This is because these are always just the filenames like in the navigation bar. The `rmd` vector may have other parts of the file path included, which is why it won't always work (we have to preserve the path to be able to read the title later on).

I think this change makes this function better, thanks.

**Comment by JiaxiangBU** [comment]
> > Next topic, I find based on this function, the toc is updated manually by users every time. Maybe an automatic solution is better.
> 
> Could you please elaborate on this idea? What would happen automatically?

For exmaple, in one of my workflowr project, I add a analysis document by `wflow_open`. And then I feel good, so use `wflow_build`, commit this adding, `wflow_publish`. I learn these steps from the tutorial you write.

However, in this processs, the toc I think is not updated. the new analysis document is not listed in the toc. Does it make sense to automatically add it in the toc?

**Comment by JiaxiangBU** [comment]
> @JiaxiangBU I just pushed two new commits to your branch. Please pull them locally with `git pull origin dev`.

John, I update this package locally by `git pull origin dev` and on the GitHub by a new pull request. Thanks for your tips.

**Comment by jdblischak** [comment]
> However, in this processs, the toc I think is not updated. the new analysis document is not listed in the toc. Does it make sense to automatically add it in the toc?

@JiaxiangBU Thanks for the explanation. I see how this would be convenient. Unfortunately this would be currently different to implement (i.e. to automatically re-run `index.Rmd` and insert an updated table of contents).

As a compromise, one option would be to add the following code chunk to `analysis/index.Rmd`:

````
```{r toc, results="asis", echo=FALSE}
suppressMessages(toc <- workflowr::wflow_toc())
cat(toc, sep = "\n")
```
````

This way, every time you built/published `index.Rmd`, it would insert an updated version of the table of contents.

**Note:** In general running workflowr functions from within an Rmd file is a bad idea, but this works for `wflow_toc()`.

**Comment by JiaxiangBU** [comment]
> ```{r toc, results="asis", echo=FALSE}
> suppressMessages(toc <- workflowr::wflow_toc())
> cat(toc, sep = "\n")
> ```

Thus, I think just print a message with the content you write for the users. Users could just paste it in their `index.Rmd`. I think it is an option for users.



> **Note:** In general running workflowr functions from within an Rmd file is a bad idea, but this works for `wflow_toc()`.

Could you please explain this idea?  Is there any limitation on the workflowr functions. I am just curious.

**Comment by jdblischak** [comment]
> Could you please explain this idea? Is there any limitation on the workflowr functions. I am just curious.

@JiaxiangBU The main workflowr functions like `wflow_build()`/`wflow_publish()`/`wflow_status()` all perform operations based on the current status of the R Markdown files in the workflowr project. Thus if you run a workflowr command from within an R Markdown file contained in a workflowr project, this can become circular. It will probably still execute, but long-term it is not a good idea for maintaining a workflowr project, and it will likely lead to problems. Does that make sense?

**Comment by jdblischak** [comment]
@JiaxiangBU I think this PR is ready to merge. Could you please add an entry to `NEWS.md`? Please reference this PR, your GitHub username, and the original Issue #151.

Here's an example of a past entry:

```
* Improve support for hosting on Shiny Server. Setting the option
`fig_path_ext: false` in `_workflowr.yml` removes the file extension from the
figure subdirectories, allowing them to be viewed on Shiny Server (implemented
by @Tutuchan, #119, #122)
```

Also, would you like to be included as an official contributor?

**Comment by JiaxiangBU** [comment]
> > Could you please explain this idea? Is there any limitation on the workflowr functions. I am just curious.
> 
> @JiaxiangBU The main workflowr functions like `wflow_build()`/`wflow_publish()`/`wflow_status()` all perform operations based on the current status of the R Markdown files in the workflowr project. Thus if you run a workflowr command from within an R Markdown file contained in a workflowr project, this can become circular. It will probably still execute, but long-term it is not a good idea for maintaining a workflowr project, and it will likely lead to problems. Does that make sense?

I get your point. Thanks for your explanation.

**Comment by JiaxiangBU** [comment]
> @JiaxiangBU I think this PR is ready to merge. Could you please add an entry to `NEWS.md`? Please reference this PR, your GitHub username, and the original Issue #151.
> 
> Here's an example of a past entry:
> 
> ```
> * Improve support for hosting on Shiny Server. Setting the option
> `fig_path_ext: false` in `_workflowr.yml` removes the file extension from the
> figure subdirectories, allowing them to be viewed on Shiny Server (implemented
> by @Tutuchan, #119, #122)
> ```

Hi, John, I make several commits to update news. I find there are several changes from origin dev, when I use `git pull origin dev`, I commit them separately, 4a88cc7.

**Comment by JiaxiangBU** [comment]
> Also, would you like to be included as an official contributor?

Yes, I do like it.

**Review by jdblischak — APPROVED** [review]

**Comment by jdblischak** [comment]
@JiaxiangBU I added you as a contributor in https://github.com/jdblischak/workflowr/commit/54b2afbadf8406d3cf7dfd4e37790f7596ecd295. The function `wflow_toc()` will be included in the next release. Thanks again for your contribution to workflowr!
